android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -83,32 +83,30 @@ class TranslationsMiddleware(
     private suspend fun requestSupportedLanguages(
         context: MiddlewareContext<BrowserState, BrowserAction>,
         tabId: String,
-    ) = withContext(Dispatchers.IO) {
-        scope.launch {
-            engine.getSupportedTranslationLanguages(
-
-                onSuccess = {
-                    context.store.dispatch(
-                        TranslationsAction.TranslateSetLanguagesAction(
-                            tabId = tabId,
-                            supportedLanguages = it,
-                        ),
-                    )
-                    logger.info("Success requesting supported languages.")
-                },
-
-                onError = {
-                    context.store.dispatch(
-                        TranslationsAction.TranslateExceptionAction(
-                            tabId = tabId,
-                            operation = TranslationOperation.FETCH_LANGUAGES,
-                            translationError = TranslationError.CouldNotLoadLanguagesError(it),
-                        ),
-                    )
-                    logger.error("Error requesting supported languages: ", it)
-                },
-            )
-        }
+    ) = withContext(MainScope().coroutineContext) {
```

Comment:
Not sure of the recommended syntax here as `scope` also happens to be `MainScope` on this file.

Reply:
Coroutine and scopes are actually my greatest unknown and weakness. So, I will defer this to @MatthewTighe and @Amejia481, but happy to just follow along and learn something in the process.

Reply:
I think the main idea of having a coroutines scope is defining until when your task should run until it can be automatically cancelled. For example, imagine that you have a task that fetches details of an item over the network when you are on a detailed screen, but what happens if the user leave the details screen, before the task completes? Ideally the task should be cancel as it's not needed anymore. That's the role of the coroutines scope, scoping the life of some coroutines.

To define the scope the question that should ask is until when this code should be run until it is cancelled,  
We have [different per-built scopes](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/) like  [LifecycleScope](https://developer.android.com/topic/libraries/architecture/coroutines#lifecyclescope) or we can create our [own custom scope depending](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-global-scope/) on our needs.

Reply:
There are other middlewares that are also receiving scopes like the [LinkingMiddleware](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/LinkingMiddleware.kt#L27)

Reply:
I think that might lend to using `= withContext(scope.coroutineContext)` then? So, it'll be tied to the scope of `TranslationsMiddleware.kt`?

Reply:
When using `withContext` ideally we would like to be explicit about which [dispatcher](https://kotlinlang.org/docs/coroutine-context-and-dispatchers.html) the code should be executed like.

`withContext(Dispatchers.IO)` or `withContext(Dispatchers.Main)`, at the moment we are just extracting it from the scope.

Reply:
it depends of the case, as you are already on suspend function, you don't need either `withContext` or `scope.launch` as you are already in the context of coroutine, see[ requestSupportedLanguages is getting from a coroutine](https://github.com/mozilla-mobile/firefox-android/blob/a83a0f49d3cb69718cf2663b12a43a59c37420c6/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt#L46). You use `withContext` or `scope.launch` when you want to specific a dispatcher that is different from the one called the suspend function.

Reply:
So you can remove `withContext(MainScope().coroutineContext)` and the code should continue working.

Reply:
Okay, that makes sense! Yes, I had tested without it at all and it works great. I guess I'm wondering if I need to be explicit about scopes so someone doesn't somehow end up [here](https://searchfox.org/mozilla-central/rev/5c4a45eb17423373ecb71aea9819d41a6231613e/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoResult.java#303) again down the road or if this is just a weird one-off odd scenario I landed myself in.

I'm also testing something with @kycn to try and keep a `Dispatcher.IO`. I think ideally we want to use that if possible.

Reply:
You could pass the dispatcher directly to `scope.launch(Dispatchers.Main)` on [requestSupportedLanguages](https://github.com/mozilla-mobile/firefox-android/blob/a83a0f49d3cb69718cf2663b12a43a59c37420c6/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt#L46), I think gecko results are expecting to be run on main, but give it a try.

Reply:
Sounds good, `scope.launch(Dispatchers.Main)` works. I set it to that since it is a bit more explicit. I set the patch to that for now and removed the `withContext`. (Open to whatever seems the best course of action you recommend here - I'm getting a bit bogged down in the nuanced differences here. Thanks, @Amejia481 for taking a look and giving all the background on the differences!)

@kycn and I had a [conversation over here](https://mozilla.slack.com/archives/G01EXNFFAVC/p1706798060455179) concurrently to this one and wanted to take a closer look tomorrow too. We're trying to salvage using `Dispatcher.IO` if it is possible too and not run into the GV issue.

Something like launching again might make sense out of `Dispatcher.IO` to give us the handler is what we were wondering:

This works as well and seems like it might be launching it correctly:
```
scope.launch{
    doSomething()
}

fun doSomething() = withContext(Dispatcher.IO){
    scope.launch {
      // Work
    }
  }
```

Thanks everyone for taking a look and the help!

Reply:
Based on the conversation on Slack, the consensus is the best option is to stay in main scope (`scope.launch{}`) here.